### PR TITLE
Patch for support numba-dppy `with device_context`

### PR DESCRIPTION
**Pull request by PokhodenkoSA**
This modifications make jit() decorator use TargetDispatcher from numba-dppy.
Changes made in IntelPython/numba#57 by @AlexanderKalistratov and @1e-to.

Patch to fix SDC integration testing (IntelPython/numba#203)

This is an initial patch from IntelPython/numba for support numba-dppy `with context`.
Code is from https://github.com/IntelPython/numba/pull/96.
History:
1. @reazulhoque created comment and set the goal here: https://github.com/IntelPython/numba/pull/96#issuecomment-775311261
2. @sklam created https://github.com/numba/numba/pull/6770 and alternative more flexible version https://github.com/numba/numba/pull/6870
3. @diptorupd created issue for review and adopt @sklam PRs in numba-dppy: https://github.com/IntelPython/numba-dppy/issues/271
4. @PokhodenkoSA created numba-dppy patch for using solution from https://github.com/numba/numba/pull/6870: https://github.com/IntelPython/numba-dppy/pull/278
Some problems with @sklam solution:
- Context does not apply to nested functions calls.
5. @sklam created solution based on https://github.com/numba/numba/pull/6814, https://github.com/numba/numba/pull/6762 and https://github.com/numba/numba/pull/6870 with introducing "legalization pass that scan for invalid target in callees" for checking nested functions calls: https://github.com/sklam/numba/blob/demo/extendtarget_demo_1/demo.md
- Contains an issue in Case 4.
- Numba tests fails on `demo` branch.

I have created this PR for sharing initial PR code which represents our initial goal. Also I have shared history of solutions we have tried to replace initial PR.

This PR is for discussion and transforming it to something appropriate for upstreaming to numba/numba.

### CODE REVIEW

**Review by sklam — COMMENTED** [review]
I have left a few comments and questions.

In addition, can you point us to intructions to run the dppy testsuite?

FYI, https://github.com/numba/numba/pull/6888 is a WIP attempt to address the "case 4" problem.

File: numba/core/dispatcher.py [hunk 1]
```diff
@@ -742,7 +742,14 @@ def _set_uuid(self, u):
         self._recent.append(self)
 
 
-class Dispatcher(serialize.ReduceMixin, _MemoMixin, _DispatcherBase):
+import abc
+
+class DispatcherMeta(abc.ABCMeta):
+    def __instancecheck__(self, other):
+        return type(type(other)) == DispatcherMeta
```
[the remark this answers is not in the document]
Comment (AlexanderKalistratov) [reply numba/core/dispatcher.py:749]: TargetDispatcher is not a subclass of Dispatcher. This is just a workaround to minimize code changes

File: numba/core/dispatcher.py [hunk 2]
```diff
@@ -995,6 +1002,9 @@ def get_function_type(self):
             cres = tuple(self.overloads.values())[0]
             return types.FunctionType(cres.signature)
 
+    def get_compiled(self):
+        return self
```
[the remark this answers is not in the document]
Comment (AlexanderKalistratov) [reply numba/core/dispatcher.py:1006]: Not sure I remember exact details. `get_compiled` is different for `TargetDispatcher`, and here is only needed for compatibility

File: numba/core/decorators.py [hunk 3]
```diff
@@ -222,7 +211,42 @@ def wrapper(func):
...
+            if target_ is None:
+                target_ = 'cpu'
+            disp = registry.dispatcher_registry[target_]
+            return wrapper(func, disp)
+
+        from numba_dppy.target_dispatcher import TargetDispatcher
+        disp = TargetDispatcher(func, wrapper, target, targetoptions.get('parallel'))
+        return disp
+
+    return __wrapper
```
[the remark this answers is not in the document]
Comment (AlexanderKalistratov) [reply numba/core/decorators.py:249]: `TargetDispatcher` is located in dppy, so there is dppy-specific code.

`We'll need to find out why we can't just use the dispatcher_registry`
Exact target is not known at this point yet.

**Comment by AlexanderKalistratov** [comment]
@sklam I don't think most of changes would be needed in upcoming solution. Except probably setting default target to `None` and extra `@global_compiler_lock`

**Comment by sklam** [comment]
> @sklam I don't think most of changes would be needed in upcoming solution. Except probably setting default target to `None` and extra `@global_compiler_lock`

Those changes make sense.

**Comment by stuartarchibald** [comment]
@PokhodenkoSA Is this patch still applicable given the history noted in the OP? If not, can the PR be closed? Thanks for your help.

**Comment by PokhodenkoSA** [comment]
I think it could be closed.

**Comment by stuartarchibald** [comment]
> I think it could be closed.

Great, thanks for confirming.

**Pull request closed without merging**